src/async_impl/decoder.rs

```diff
@@ -0,0 +1,300 @@
+{
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        // Do a read or poll for a pendidng decoder value.
+        let new_value = match self.inner {
+            DecoderInner::Pending(ref mut future) => {
+                match future.poll() {
+                    Ok(Async::Ready(inner)) => inner,
+                    Ok(Async::NotReady) => return Err(io::ErrorKind::WouldBlock.into()),
+                    Err(e) => return Err(error::into_io(e))
+                }
+            },
+            DecoderInner::PlainText(ref mut body) => return body.read(buf),
+            DecoderInner::Gzip(ref mut decoder) => return decoder.read(buf)
+        };
+
+        self.inner = new_value;
+        self.read(buf)
+    }
+}
+
+impl<TStream> Pending<TStream> {
+    fn into_inner(self) -> TStream {
+        match self {
+            Pending::Gzip(reader) => reader.into_inner(),
+            Pending::Empty => panic!("attempted to take used reader")
+        }
+    }
+}
+
+impl<TStream> Future for Pending<TStream>
+    where TStream: Stream<Item = Chunk, Error = error::Error>
+{
+    type Item = DecoderInner<TStream>;
+    type Error = error::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let body_state = match *self {
+            Pending::Gzip(ref mut body) => {
+                match body.poll_stream() {
+                    Ok(Async::Ready(state)) => state,
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Err(e) => return Err(e)
+                }
+            },
+            Pending::Empty => panic!("poll for a decoder after it's done")
+        };
+
+        match mem::replace(self, Pending::Empty) {
+            Pending::Gzip(body) => {
+                // libflate does a read_exact([0; 2]), so its impossible to tell
+                // if the stream was empty, or truly had an UnexpectedEof.
+                // Therefore, we need to check for EOF first.
+                match body_state {
+                    StreamState::Eof => Ok(Async::Ready(DecoderInner::PlainText(body))),
+                    StreamState::HasMore => Ok(Async::Ready(DecoderInner::Gzip(gzip::Decoder::new(body))))
+                }
+            },
+            Pending::Empty => panic!("invalid internal state")
+        }
+    }
+}
+
+/// A `Read`able wrapper over a stream of chunks.
+struct ReadableChunks<TStream> {
+    state: ReadState,
+    stream: TStream,
+}
+
+enum ReadState {
+    Ready(Chunk, usize),
+    NotReady,
+    Eof,
+}
+
+enum StreamState {
+    HasMore,
+    Eof
+}
+
+impl<TStream> ReadableChunks<TStream> {
+    #[inline]
+    fn new(stream: TStream) -> Self {
+        ReadableChunks {
+            state: ReadState::NotReady,
+            stream: stream,
+        }
+    }
+
+    fn into_inner(self) -> TStream {
+        match self.state {
+            ReadState::Ready(_, _) => panic!("attempted to take used reader"),
```

Comment:
We might want to rethink these error conditions. They should effectively be unreachable, but the invariants are spread a bit thin.

Maybe it would be better to return `Option` or `Result` here and `expect` it in the synchronous response? Or just silently ignore the first chunk of there is one, but that seems worse to me.